app/routes/login.js

```diff
@@ -67,6 +68,12 @@ export default Route.extend({
       let user = this.store.push(this.store.normalize('user', data.user));
       let transition = this.get('session.savedTransition');
       this.session.loginUser(user);
+
+      const messageAfterLogin = htmlSafe(
+        "Welcome to crates.io! Visit <a href='me'>account settings</a> to verify your email address and create an API token!",
+      );
+      this.flashMessages.show(messageAfterLogin, { type: 'info' });
```

Comment:
do you want to show this message to **all** users? even if they have verified their data already? 🤔

Reply:
For first version yes. Currently trying  to understand how I can pass data from backend and what parameters it will be. I think about `{ "email_verified": false, "api_key_verified": true }` but backed currently is a real mess for me. No good structure and abstractions after rails. Still trying to learn how to handle that.

Reply:
Hi @mordorreal! Thank you for this pull request!

I'm sorry our backend is such a mess; there isn't really a Rails for Rust yet and crates.io was written even before any of the experimental web frameworks existed!

There is a [router.rs file](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs) that is similar (in purpose at least) to Rails' routes.rb. There are two backend routes that will be relevant for this PR:

- [`/api/private/session/authorize`](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L118) is how the frontend gets information about the current user from the backend immediately after logging in with GitHub. It [calls](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L119) the [`user::session::authorize` controller function](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/controllers/user/session.rs#L46-L105). After interacting with info from GitHub, it [calls](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/controllers/user/session.rs#L104) the `user::me::me` controller function (discussed more next)
- [`/api/v1/me`](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/router.rs#L79) is how the frontend gets information about the current user from the backend if there's already an existing session for a logged in user. It also calls the [`user::me::me` controller function](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/controllers/user/me.rs#L14-L61), so this URL returns the same data as the session authorize URL.

The data returned currently is defined by the "view" struct [`EncodableMe`](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L159-L163). Because the backend is only a JSON API, there are no HTML views, so all the views are structs that get serialized to JSON with serde.

`EncodableMe` contains an [`EncodablePrivateUser`](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L165-L178), which has an [`email_verified` field](https://github.com/rust-lang/crates.io/blob/f82b3926e4fbe787228d462b9aef0a8b734c9494/src/views.rs#L172), so you should be able to use that!

We can add a field to `EncodablePrivateUser` called `has_tokens` or similar that is `true` if the user has at least one token and `false` if they have none. We'd need to add a query to `user::me::me` that gets that information. The query should use `ApiToken::belonging_to` sort of like [this code does](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/controllers/token.rs#L13-L14) and diesel's `exists` function sort of like [this code does](https://github.com/rust-lang/crates.io/blob/4b1aa086e702bfd741773b5265cf149fe2c6a6b7/src/models/krate.rs#L163-L169). I'm actually not sure off the top of my head exactly how to combine those two! [Diesel's docs](http://docs.diesel.rs/) is where I'd look to figure that out. Want to give that a try? 

Let me know if you have more questions, I hope that helps to pick the relevant pieces out of our mess at least!

Reply:
Hi @carols10cents! Thank you for very detailed answer! You're awesome! I'll try to continue on the issue in this PR.